### DISCUSSION TITLE
Podman "database cannot be opened"

### DISCUSSION THREAD

**Discussion by espentrydal** [abraunegg/onedrive#2524] [Q&A] [answered]
Hi

I am running Fedora 38 but I have had the same problem at another computer running Ubuntu, so probably I am doing something wrong.

I get this error message after pasting the response uri:

```
ERROR: The local file system returned an error with the following message:
  Error Message:    /onedrive/conf/refresh_token: Permission denied

The database cannot be opened. Please check the permissions of ~/.config/onedrive/items.sqlite3

/entrypoint.sh: line 125:    10 Segmentation fault      (core dumped) /usr/local/bin/onedrive "${ARGS[@]}"
usermod: no changes
Base Args: --monitor
```
The folder .config/onedrive exists with normal user permissions but empty (items.sqlite3 does not exist).

My installation script as well as I managed to understand it from the onedrive podman documentation:

```
export ONEDRIVE_DATA_DIR="${HOME}/OneDrive"
export ONEDRIVE_UID=1000
export ONEDRIVE_GID=1000
mkdir -p ${ONEDRIVE_DATA_DIR}

podman run -it --name onedrive --user "${ONEDRIVE_UID}:${ONEDRIVE_GID}" \
        --userns=keep-id \
        -v onedrive_conf:/onedrive/conf:U,Z \
        -v "${ONEDRIVE_DATA_DIR}:/onedrive/data:U,Z" \
        -e PODMAN=1 \
        --label "io.containers.autoupdate=image" \
        docker.io/driveone/onedrive:latest
```

(Without userns I got an error as mentioned in the readme. 

I did struggle sometimes to understand what was meant in some parts of the readme. For example, under step 1.1 I assume it is the second and not the first volume, and under step 2 it feels like the second "Important" paragraph mentioning --userns is inserted between two paragraphs that should be together).

**Comment by abraunegg** [comment]
@espentrydal 

You are basically hitting permission issues - which is most likely because the correct steps were not followed.

> For example, under step 1.1 I assume it is the second and not the first volume,

It looks like the wording is messed up and needs to be corrected - potentially when podman and docker documentation was being aligned, this got messed up.

2 volumes are needed:
* Config Volume
* Data Volume

The first is the configuration volume that stores all the applicable application configuration + runtime state. In a non-containerised environment, this normally resides in `~/.config/onedrive` - in a containerised environment this is stored in the volume tagged as `/onedrive/conf`

The second volume is the data volume, where all your data from Microsoft OneDrive is stored locally. This volume is mapped to an actual directory point on your local filesystem and this gets mapped to `/onedrive/data`

Your specific issue is that Podman cannot write to this configuration volume which resides inside the container (for example writing `/onedrive/conf/refresh_token`) .. so this 100% is a permission issue on how you have setup and configured Podman in your environment.


> and under step 2 it feels like the second "Important" paragraph mentioning --userns is inserted between two paragraphs that should be together

Yes that does appear to be in the wrong spot + is missing the example that goes along with it.

The documentation will be updated as part of 'alpha-3' testing.

> `docker.io/driveone/onedrive:latest`

The 'latest' is not actually 'latest' ... it is the latest actual 'release'. There have been 11 commits since the last release, some fixing some bugs, others are documentation updates, it would be advisable for you use `onedrive:edge` rather than `onedrive:latest`. Please review https://github.com/abraunegg/onedrive/compare/v2.4.25...master for details

**Comment by espentrydal** [comment]
Thank you @abraunegg for your swift reply.

The steps I followed are as in the script I attached above. I tried again with explicitly setting up the config volume first with `podman volume create onedrive_conf`, but I get the same error. As root or user makes no difference. I also tried with :edge and :edge-fedora, but no change.

My Fedora install was a fresh install of 38, so I don't have any particular podman configuration except what comes with the install as far as I am aware (I have not used podman for anything else on this machine). My uid and gid are both 1000. I've disabled selinux. I've reinstalled podman.

Unless there is something I've overlooked which still is very possible, one possiblity is that the problem was introduced either by an update of the image, or less likely, by a more recent version of podman. The last time I managed to get it to work was on Ubuntu 22.04 LTS, probably with an image more than 4 months old, which is older than what is available on docker.io now.

I am happy to try out something else if you have any suggestions, or maybe an address for an older image if that is possible.

On Fedora I now realize that it works well just as a normal app, this was not the case for Ubuntu which is why I used the container version. So for now I am good with that solution on the Fedora machine. Though it would be great to make it work on the Ubuntu machine too.

**Comment by abraunegg** [comment]
@espentrydal 
I have been able to replicate your issue on a brand new Fedora 38 install.

As per the documentation (yes, whilst the other parts you have called out need better clarity), the following is quite clear:

![image](https://github.com/abraunegg/onedrive/assets/4956234/ab2d1cc8-60bb-4fbc-ad5e-73512de8f2e3)

With SELinux enabled, this is the error I get:
```
usermod: no changes
Base Args: --monitor
# Launching onedrive
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=d50ca740-c83f-4d1b-b616-12c519384f0c&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<redacted>

ERROR: The local file system returned an error with the following message:
  Error Message:    /onedrive/conf/refresh_token: Permission denied

The database cannot be opened. Please check the permissions of ~/.config/onedrive/items.sqlite3

/entrypoint.sh: line 132:    10 Segmentation fault      (core dumped) /usr/local/bin/onedrive "${ARGS[@]}"
```

When SELinux is disabled correctly, no such error occurs.

Please double check that SELinux has been correctly disabled on your system.

**Comment by espentrydal** [comment]
Thank you @abraunegg for checking. 

I understand it must be on my side, but I have a hard time finding what it could be. On Fedora 38 SELinux is disabled as far as I can tell:

`cat /proc/cmdline`: 
`BOOT_IMAGE=(hd0,gpt2)/vmlinuz-6.5.7-200.fc38.x86_64 [...] selinux=0`

`sestatus`:
`SELinux status:                 disabled`

And I still get the same error message. The first times I did try to run it with SELinux enabled or in permissive mode, though since that I have reinstalled podman with SELinux disabled. I guess it is likely something is still sticking though I don't understand what it could be.

On Ubuntu 22.04 I  also still get the exact same error message. There SELinux is not installed and I have uninstalled AppArmor, as well as told the kernel not to load it.

`cat /proc/cmdline`: 
`BOOT_IMAGE=vmlinuz-6.2.0/generic [...] apparmor=0`

**Comment by abraunegg** [comment]
Potentially start from scratch - delete your volumes and container and start again.

Im also updating the doco for alpha-3 - will post that once done

**Comment by espentrydal** [comment]
Thanks. Yes, I have done that every time, if not the commands fail (`podman container rm onedrive; podman volume rm onedrive_conf`). The Ubuntu is a multiuser machine so unfortunately I cannot experiment too much (and I should reinstall apparmor). I'll try any of the other methods at a later time. On the Fedora the simple install worked well.

**Comment by abraunegg** [comment]
@espentrydal 
Please can you review the updated podman documentation as detailed here: https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-alpha-3/docs/podman.md

**Comment by espentrydal** [comment]
@abraunegg 
Very clear instructions, easy to follow. Unfortunately I still get the same error message both on Fedora (selinux disabled, non-containerized onedrive stopped and disabled and existing onedrive-config moved away) and Ubuntu.

```
Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<...>

ERROR: The local file system returned an error with the following message:
  Error Message:    /onedrive/conf/refresh_token: Permission denied

The database cannot be opened. Please check the permissions of ~/.config/onedrive/items.sqlite3

/entrypoint.sh: line 125:    10 Segmentation fault      (core dumped) /usr/local/bin/onedrive "${ARGS[@]}"
```

I have successfully installed the OpenSuSE build service version so I now have onedrive working on both the Ubuntu machine and my Fedora machine. But I am availalbe to test the podman version further. Thank you for a very useful client and your help.

**Comment by abraunegg** [comment]
> Unfortunately I still get the same error message both on Fedora (selinux disabled, non-containerized onedrive stopped and disabled and existing onedrive-config moved away) and Ubuntu.

It has to be asked - how are you disabling SELinux ? Are you following *this* process - https://docs.fedoraproject.org/en-US/quick-docs/selinux-changing-states-and-modes/#_disabling_selinux

![image](https://github.com/abraunegg/onedrive/assets/4956234/5badebb3-7385-4cea-b2b4-50d29a11a964)

This is the procedure that I use. This *should* be the *same* procedure you are using. The procedure should be the same for Fedora and Ubuntu platforms.

**Reply by espentrydal** [reply]
I understand. Yes. Additionally I do `grubby --update-kernel ALL --args selinux=0` to fully disable it in the kernel just in case (this seems to be a new requirement in 38). 

On Ubuntu selinux is not installed. I uninstalled AppArmor during some of the tests but that didn't change anything.

**Reply by abraunegg** [reply]
> Additionally I do grubby --update-kernel ALL --args selinux=0 to fully disable it in the kernel just in case (this seems to be a new requirement in 38).

I do not think this is accurate. The updated podman document was written using Fedora 38.

Perhaps remove that, ensure it is disabled correctly, reboot, check via `getenforce` .. there *has* to be something on your system that is causing this issue.

**Reply by abraunegg** [reply]
@espentrydal 
My suggest is the following:
1. Remove your `grubby` commands and boot loader attempt to disable SELinux
2. Configure the SELinux config file as per above
3. Tell your system to relabel your filesystem: `touch /.autorelabel`
4. Reboot

![image](https://github.com/abraunegg/onedrive/assets/4956234/a12469d6-ee29-42bf-93e7-cdc5a49553b9)

**Comment by espentrydal** [comment]
So I did as instructed. Same error message. For info:

`getenforce`
`Disabled`

`cat /proc/cmdline`
`BOOT_IMAGE=(hd0,gpt2)/vmlinuz-6.5.7-200.fc38.x86_64 root=UUID=6a224e0d-ccd7-43f4-bd66-02ae2fbdc2f8 ro rootflags=subvol=root rd.driver.blacklist=nouveau modprobe.blacklist=nouveau nvidia-drm.modeset=1 rd.luks.uuid=luks-398689b7-5484-4ca3-8cdf-ba69f4fdfca9 rhgb quiet rd.driver.blacklist=nouveau modprobe.blacklist=nouveau nvidia-drm.modeset=1
`
(no selinux=0)

```
ls -a /                                                                                                                                          
.  ..  afs  .autorelabel  bin  boot  dev  etc  home  lib  lib64  lost+found  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
```

Why I also did selinux=0:

 `/etc/selinux/config`
```
# NOTE: In earlier Fedora kernel builds, SELINUX=disabled would also
# fully disable SELinux during boot. If you need a system with SELinux
# fully disabled instead of SELinux running with no policy loaded, you
# need to pass selinux=0 to the kernel command line. You can use grubby
# to persistently set the bootloader to boot with selinux=0:
#
#    grubby --update-kernel ALL --args selinux=0
#
# To revert back to SELinux enabled:
#
#    grubby --update-kernel ALL --remove-args selinux
```
(It's late here so I am not able to test further today)

**Reply by abraunegg** [reply]
Did you relabel your filesystem ?

**Comment by espentrydal** [comment]
@abraunegg My understanding is that the relabeling only happens when SELinux is enabled, but I both have the `.autorelabel` file in root and I did run the `sudo fixfiles onboot` to be sure (I have only observed the actual relabeling happening when I have turned SELinux on after having it disabled).

**Reply by espentrydal** [reply]
To be sure I also followed the instructions here to manually remove SELinux labels, but it had no effect: [Getting rid of SELinux labels](https://stephane.lesimple.fr/blog/getting-rid-of-selinux-labels/)

**Comment by abraunegg** [comment]
@espentrydal 
Unfortunately, I am out of options here. The issue you are facing I cannot replicate. The instructions as provided were written whilst using Fedora 38 & the Fedora 38 provided version of podman - thus are known to work.

There is something *unique* regarding your environment and/or setup. I suggest that you take your podman issue to their discussion forums: https://podman.io/community

**Comment by espentrydal** [comment, marked as answer]
@abraunegg 
Thanks for your help. Follwing the first steps here [Container permission denied: How to diagnose this error](https://www.redhat.com/sysadmin/container-permission-denied-errors) I found one way to make the container run, by appending `--cap-add=DAC_OVERRIDE` to the `podman run` command. So it seems to be a file permission issue (cf. `man capabilities`).

Docker works fine, which makes sense since it runs under more extensive permissions than podman.

**Reply by espentrydal** [reply]
Just a shot in the dark as it is almost certanly my config which is the issue, but just in case, I can see that the `/onedrive` folder is owned by root. Could that be an issue when spinning up the image? `/onedrive/conf` and `/ondrive/conf/refresh_token` are owned by my user with user rw permissions for the file (after having started the container successfully with the above capability, and then entering the container with `podman exec -it onedrive /bin/bash`).

**Reply by abraunegg** [reply]
Nothing should be owned by root at all .. so again  this has to be something with your config or environment.

**Reply by espentrydal** [reply]
Ok. Well, at least I have a workaround if it becomes necessary.
